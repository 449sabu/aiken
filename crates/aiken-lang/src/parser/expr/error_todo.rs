use chumsky::prelude::*;

use crate::{
    expr::UntypedExpr,
    parser::{
        error::ParseError,
        expr::{block::parser as block, string},
        token::Token,
    },
};

pub fn parser(
    sequence: Recursive<'_, Token, UntypedExpr, ParseError>,
) -> impl Parser<Token, UntypedExpr, Error = ParseError> + '_ {
    let message = || choice((string::hybrid(), block(sequence.clone())));
    choice((
        just(Token::Todo)
            .ignore_then(message().or_not())
            .map_with_span(UntypedExpr::todo),
        just(Token::ErrorTerm)
            .ignore_then(message().or_not())
            .map_with_span(UntypedExpr::error),
    ))
}

#[cfg(test)]
mod tests {
    use crate::assert_expr;

    #[test]
    fn error_basic() {
        assert_expr!(
            r#"
            error @"foo"
            "#
        );
    }

    #[test]
    fn error_sugar() {
        assert_expr!(
            r#"
            error "foo"
            "#
        );
    }

    #[test]
    fn todo_basic() {
        assert_expr!(
            r#"
            todo @"foo"
            "#
        );
    }

    #[test]
    fn todo_sugar() {
        assert_expr!(
            r#"
            todo "foo"
            "#
        );
    }
}
