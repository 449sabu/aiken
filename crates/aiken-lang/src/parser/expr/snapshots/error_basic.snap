---
source: crates/aiken-lang/src/parser/expr/error_todo.rs
description: "Code:\n\nerror @\"foo\"\n"
---
Trace {
    kind: Error,
    location: 0..12,
    then: ErrorTerm {
        location: 0..12,
    },
    text: String {
        location: 6..12,
        value: "foo",
    },
}
